### PR TITLE
Fixes #56 : new add functions and test for the changes

### LINKED ISSUE #56: Test correct calculation of the `res` auxiliar value

### PR DESCRIPTION
**Pull request by jkktom**
Fixes #56
I made changes to the memory_value.go and vm_test.go
Please review and let me know the followup actions.

### CODE REVIEW

File: pkg/vm/memory/memory_value.go [hunk 1]
```diff
@@ -194,19 +183,65 @@ func (mv *MemoryValue) Equal(other *MemoryValue) bool {
...
 		if rhs.IsAddress() {
-			mv.address, err = mv.address.Add(rhs.address, lhs.felt)
-		} else {
+			// lhs : felt, rhs : Address
+			return nil, errors.New("invalid operation: cannot add integer to memory address")
+		} else if rhs.IsFelt() {
+			// lhs : felt, rhs : felt
 			mv.felt = mv.felt.Add(lhs.felt, rhs.felt)
+		} else {
+			return nil, errors.New("invalid rhs type")
```
Comment (rodrigo-pino) [inline pkg/vm/memory/memory_value.go:244]: It is safe to go for just an else here, no need to check if it is a felt. Memory values are either one or the other. There are never the two at the same time, or none.

File: pkg/vm/memory/memory_value.go [hunk 2]
```diff
@@ -194,19 +183,65 @@ func (mv *MemoryValue) Equal(other *MemoryValue) bool {
...
+				return nil, errors.New("cannot add addresses from different segments")
+			}
+			mv.address.SegmentIndex = lhs.address.SegmentIndex
+			mv.address.Offset = lhs.address.Offset + rhs.address.Offset
+		} else if rhs.IsFelt() {
+			// lhs : Address, rhs : felt
+			mv.address, err = mv.address.Add(lhs.address, rhs.felt)
+		} else {
+			return nil, errors.New("invalid rhs type")
 		}
```
Comment (rodrigo-pino) [inline pkg/vm/memory/memory_value.go:235]: A simple `else` is enough here, no need to check if it is a field element after checking if it is an address

File: pkg/vm/memory/memory_value.go [hunk 3]
```diff
@@ -194,19 +183,65 @@ func (mv *MemoryValue) Equal(other *MemoryValue) bool {
...
+// Adds a memory address and a field element
+func (address *MemoryAddress) Add(lhs *MemoryAddress, rhs *f.Element) (*MemoryAddress, error) {
+	if !rhs.IsUint64() {
+		return nil, fmt.Errorf("field element does not fit in uint64: %s", rhs.String())
+	}
+
+	address.SegmentIndex = lhs.SegmentIndex
+	address.Offset = lhs.Offset + rhs.Uint64()
+	return address, nil
+}
```
Comment (rodrigo-pino) [inline pkg/vm/memory/memory_value.go:218]: Why did you change the position of this function definition?

File: pkg/vm/memory/memory_value.go [hunk 4]
```diff
@@ -194,19 +183,65 @@ func (mv *MemoryValue) Equal(other *MemoryValue) bool {
...
+// 		} else {
+// 			mv.felt = mv.felt.Add(lhs.felt, rhs.felt)
+// 		}
+// 	}
+
+// 	if err != nil {
+// 		return nil, err
+// 	}
+// 	return mv, nil
+// }
```
Comment (rodrigo-pino) [inline pkg/vm/memory/memory_value.go:207]: Why are we leaving this comment here?

File: pkg/vm/vm_test.go [hunk 5]
```diff
@@ -330,6 +342,76 @@ func TestComputeAddRes(t *testing.T) {
...
+			name:     "Type A : MemoryAddressPlusInteger",
+			op0:      mem.MemoryValueFromMemoryAddress(mem.NewMemoryAddress(2, 10)),
+			op1:      mem.MemoryValueFromInt(15),
+			expected: mem.MemoryValueFromMemoryAddress(mem.NewMemoryAddress(2, 25)),
+		},
+		{
+			name:     "Type B : MemoryAddressPlusMemoryAddress_SameSegment",
+			op0:      mem.MemoryValueFromMemoryAddress(mem.NewMemoryAddress(2, 10)),
+			op1:      mem.MemoryValueFromMemoryAddress(mem.NewMemoryAddress(2, 15)),
+			expected: mem.MemoryValueFromMemoryAddress(mem.NewMemoryAddress(2, 25)),
```
Comment (rodrigo-pino) [inline pkg/vm/vm_test.go:364]: This should crash, since address + address is not allowed

File: pkg/vm/vm_test.go [hunk 6]
```diff
@@ -330,6 +342,76 @@ func TestComputeAddRes(t *testing.T) {
...
+			name:     "Type D : IntegerPlusInteger",
+			op0:      mem.MemoryValueFromInt(10),
+			op1:      mem.MemoryValueFromInt(15),
+			expected: mem.MemoryValueFromInt(25),
+		},
+		{
+			name:       "Type E : IntegerPlusMemoryAddress",
+			op0:        mem.MemoryValueFromInt(15),
+			op1:        mem.MemoryValueFromMemoryAddress(mem.NewMemoryAddress(2, 10)),
+			shouldFail: true,
```
Comment (rodrigo-pino) [inline pkg/vm/vm_test.go:382]: This one should succeed since sum operations are commutative, addr + felt = felt + addr

File: pkg/vm/vm_test.go [hunk 7]
```diff
@@ -330,6 +342,76 @@ func TestComputeAddRes(t *testing.T) {
 	assert.Equal(t, expected, res)
 }
 
+func TestComputeAddRes_Jake(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jkktom) [reply pkg/vm/vm_test.go:345]: Okay. I will change it as the current style.

File: pkg/vm/memory/memory_value.go (hunk 1 above)
Comment (jkktom) [reply pkg/vm/memory/memory_value.go:244]: Got it.

File: pkg/vm/memory/memory_value.go (hunk 2 above)
Comment (jkktom) [reply pkg/vm/memory/memory_value.go:235]: Okay.

File: pkg/vm/memory/memory_value.go (hunk 3 above)
Comment (jkktom) [reply pkg/vm/memory/memory_value.go:218]: I will change it back to where it was. I was carefully checking if there's any missing cases.

File: pkg/vm/memory/memory_value.go (hunk 4 above)
Comment (jkktom) [reply pkg/vm/memory/memory_value.go:207]: I will erase. I tried to compare it side by side before / after the change

File: pkg/vm/vm_test.go (hunk 5 above)
Comment (jkktom) [reply pkg/vm/vm_test.go:364]: Okay. Will modify it.

File: pkg/vm/vm_test.go (hunk 6 above)
Comment (jkktom) [reply pkg/vm/vm_test.go:382]: Okay will look into it.
